File: backend/src/contracts/teacher-load.ts
import type { RuleSetSnapshot } from "./rule-set";

export const TEACHER_LOAD_CONTRACT_VERSION = "TEACHER-LOAD-1.0.0" as const;
export const TEACHER_NORM_RULE_CODE = "RULE-TEACH-002" as const;
export const TEACHER_NORM_WEEKS_RULE_CODE = "RULE-TEACH-003" as const;
export const TEACHER_REDUCTION_RULE_PREFIX = "RULE-TEACH-REDUCTION-" as const;

export type TeacherSchoolLevel = "THCS" | "THPT" | "THCS_THPT";
export type TeacherLoadStatus = "UNDER_TARGET" | "AT_TARGET" | "OVER_TARGET";
export type TeacherLoadEnforcement = "REPORT_ONLY" | "HARD_CAP";

export interface TeacherLoadInput {
  schoolId: string;
  academicPeriodId: string;
  teacherId: string;
  teacherCode: string;
  teacherName: string;
  schoolLevel: TeacherSchoolLevel;
  assignedWeeklySessions: number;
  asOf: string;
}

export interface TeacherLoadRuleSource {
  code: string;
  sourceUrl: string;
  sourceLocator?: string;
  ruleSetVersion: string;
  snapshotHash: string;
}

export interface TeacherLoadReduction {
  code: string;
  roleCode: string;
  reductionSessionsPerWeek: number;
  source: TeacherLoadRuleSource;
}

export interface TeacherLoadCalculation {
  contractVersion: typeof TEACHER_LOAD_CONTRACT_VERSION;
  schoolId: string;
  academicPeriodId: string;
  teacherId: string;
  teacherCode: string;
  teacherName: string;
  schoolLevel: TeacherSchoolLevel;
  weeklyNormSessions: number;
  weeklyReductionSessions: number;
  targetAverageWeeklySessions: number;
  assignedAverageWeeklySessions: number;
  teachingWeeksForNorm: number;
  annualNormSessions: number;
  annualReductionSessions: number;
  annualTargetSessions: number;
  annualAssignedSessions: number;
  weeklyVarianceSessions: number;
  status: TeacherLoadStatus;
  enforcement: TeacherLoadEnforcement;
  hardWeeklyLimitSessions: number | null;
  reductions: TeacherLoadReduction[];
  ruleSources: TeacherLoadRuleSource[];
  warnings: string[];
}

export interface TeacherLoadReport {
  contractVersion: typeof TEACHER_LOAD_CONTRACT_VERSION;
  schoolId: string;
  academicPeriodId: string;
  effectiveAsOf: string;
  ruleSnapshotId: string;
  ruleSetVersion: string;
  ruleSnapshotHash: string;
  loads: TeacherLoadCalculation[];
}

export function getTeacherLoadRuleSource(snapshot: RuleSetSnapshot, code: string): TeacherLoadRuleSource {
  const rule = snapshot.rules.find((candidate) => candidate.code === code);
  if (!rule) throw new Error(`Rule ${code} is not present in the snapshot`);
  return {
    code: rule.code,
    sourceUrl: rule.sourceUrl,
    ...(rule.sourceLocator ? { sourceLocator: rule.sourceLocator } : {}),
    ruleSetVersion: snapshot.ruleSetVersion,
    snapshotHash: snapshot.snapshotHash,
  };
}

File: backend/solver/src/timetable_solver/teacher_load.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

TEACHER_LOAD_CONTRACT_VERSION = "TEACHER-LOAD-1.0.0"


class TeacherLoadRuleSource(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(min_length=1)
    sourceUrl: str = Field(min_length=1)
    sourceLocator: str | None = Field(default=None, min_length=1)
    ruleSetVersion: str = Field(pattern=r"^RULE-SET-[0-9]+\.[0-9]+\.[0-9]+$")
    snapshotHash: str = Field(pattern=r"^[0-9a-f]{64}$")


class TeacherLoadReduction(BaseModel):
    model_config = ConfigDict(extra="forbid")

    code: str = Field(min_length=1)
    roleCode: str = Field(min_length=1)
    reductionSessionsPerWeek: float = Field(ge=0)
    source: TeacherLoadRuleSource


class TeacherLoadCalculation(BaseModel):
    model_config = ConfigDict(extra="forbid")

    contractVersion: Literal["TEACHER-LOAD-1.0.0"]
    schoolId: str = Field(min_length=1)
    academicPeriodId: str = Field(min_length=1)
    teacherId: str = Field(min_length=1)
    teacherCode: str = Field(min_length=1)
    teacherName: str = Field(min_length=1)
    schoolLevel: Literal["THCS", "THPT", "THCS_THPT"]
    weeklyNormSessions: float = Field(gt=0)
    weeklyReductionSessions: float = Field(ge=0)
    targetAverageWeeklySessions: float = Field(ge=0)
    assignedAverageWeeklySessions: float = Field(ge=0)
    teachingWeeksForNorm: int = Field(ge=1)
    annualNormSessions: float = Field(gt=0)
    annualReductionSessions: float = Field(ge=0)
    annualTargetSessions: float = Field(ge=0)
    annualAssignedSessions: float = Field(ge=0)
    weeklyVarianceSessions: float
    status: Literal["UNDER_TARGET", "AT_TARGET", "OVER_TARGET"]
    enforcement: Literal["REPORT_ONLY", "HARD_CAP"]
    hardWeeklyLimitSessions: float | None = Field(default=None, ge=0)
    reductions: list[TeacherLoadReduction]
    ruleSources: list[TeacherLoadRuleSource]
    warnings: list[str]
